crates/sui-core/src/consensus_adapter.rs

```diff
@@ -819,13 +819,15 @@ impl ConsensusAdapter {
         let send_end_of_publish = if let ConsensusTransactionKind::UserTransaction(_cert) =
             &transaction.kind
         {
-            let reconfig_guard = epoch_store.get_reconfig_state_read_lock_guard();
             // If we are in RejectUserCerts state and we just drained the list we need to
             // send EndOfPublish to signal other validators that we are not submitting more certificates to the epoch.
             // Note that there could be a race condition here where we enter this check in RejectAllCerts state.
             // In that case we don't need to send EndOfPublish because condition to enter
             // RejectAllCerts is when 2f+1 other validators already sequenced their EndOfPublish message.
-            if reconfig_guard.is_reject_user_certs() {
+            if epoch_store
```

Comment:
This is to ensure that we don't hold the lock unnecessarily long.

Reply:
I think this is safe, but it shouldn't affect anything, right? previously we held the lock until line 848, now we release it immediately because its in a temporary var, but the code below is sync and non-blocking.

Reply:
Yes, the code before was also correct, and this change basically ensures that the lock is held only when necessary. It is consistent with other places where a read lock is acquired.

The change also ensures that if we added something between here and line 848 in the future, we don't have a surprise that we are holding a lock.

Reply:
Yes, this seem safe.

After reading this code a bit more I realized that we can actually send `EndOfPublish` multiple times due to race condition on this code(multiple tasks can enter this code path, delete certs and both read that `pending_count=0`). This is ~ok, but since we are touching this code, perhaps it worth adding a comment here mentioning that this can happen.

Reply:
Sorry that the PR was auto merged... Let me add the comment in a separate PR.